scheduling-manager/src/main/java/com/amazonaws/blox/scheduling/scheduler/SchedulerHandler.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://aws.amazon.com/apache2.0/
+ *
+ * or in the "LICENSE" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+package com.amazonaws.blox.scheduling.scheduler;
+
+import com.amazonaws.blox.dataservicemodel.v1.client.DataService;
+import com.amazonaws.blox.dataservicemodel.v1.model.EnvironmentRevision;
+import com.amazonaws.blox.dataservicemodel.v1.model.operations.DescribeEnvironmentRevisionResponse;
+import com.amazonaws.blox.dataservicemodel.v1.model.operations.DescribeTargetEnvironmentRevisionRequest;
+import com.amazonaws.blox.scheduling.scheduler.engine.Scheduler;
+import com.amazonaws.blox.scheduling.scheduler.engine.SchedulerFactory;
+import com.amazonaws.blox.scheduling.scheduler.engine.SchedulingAction;
+import com.amazonaws.services.lambda.runtime.Context;
+import com.amazonaws.services.lambda.runtime.RequestHandler;
+import com.spotify.futures.CompletableFutures;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import software.amazon.awssdk.services.ecs.ECSAsyncClient;
+
+@Component
+@RequiredArgsConstructor(onConstructor = @__({@Autowired}))
+@Log4j2
+public class SchedulerHandler implements RequestHandler<SchedulerInput, SchedulerOutput> {
+  private final DataService data;
+  private final ECSAsyncClient ecs;
+  private final SchedulerFactory schedulerFactory;
+
+  @Override
+  public SchedulerOutput handleRequest(SchedulerInput input, Context context) {
+    log.debug("Request: {}", input);
+
+    DescribeEnvironmentRevisionResponse r =
```

Comment:
I usually use finals for all variables

Reply:
I was meaning to ask you about that. If I recall correctly, after Java 5 or 6 the compiler became smart enough to correctly infer `final` from context, and the only technical reason to use `final` for locals is to allow the variables to be captured in an anonymous function. Since then I stopped using `final` for locals, since it makes things noisy and harder to read, unless it's a *really* long method and there's some risk of the variable being modified in the body somewhere.

Reply:
Can you point me to the compiler context referral? Quick googling didn't surface it

Reply:
Yeah, it's kinda buried in the JLS:

From [JLS 4.12.4](http://docs.oracle.com/javase/specs/jls/se8/html/jls-4.html#jls-4.12.4):
> A local variable whose declarator has an initializer (§14.4.2) is effectively final if all of the following are true:
> * It is not declared final.
> * It never occurs as the left hand side in an assignment expression (§15.26). (Note that the local variable declarator containing the initializer is not an assignment expression.)
> * It never occurs as the operand of a prefix or postfix increment or decrement operator (§15.14, §15.15).